src/optimizer.cpp

```diff
@@ -124,67 +103,95 @@ geometry_msgs::msg::TwistStamped Optimizer::evalControl(
   const geometry_msgs::msg::Twist & robot_speed,
   const nav_msgs::msg::Path & plan, nav2_core::GoalChecker * goal_checker)
 {
-  for (size_t i = 0; i < settings_.iteration_count_; ++i) {
-    generated_trajectories_ =
-      generateNoisedTrajectories(robot_pose, robot_speed);
-    auto && costs = critic_manager_.evalTrajectoriesScores(
-      state_, generated_trajectories_, plan, robot_pose, goal_checker);
-    updateControlSequence(costs);
+  prepare(robot_pose, robot_speed, plan, goal_checker);
+
+  for (size_t i = 0; i < settings_.iteration_count; ++i) {
+    generated_trajectories_ = generateNoisedTrajectories();
+    critic_manager_.evalTrajectoriesScores(critics_data_);
+    updateControlSequence();
+  }
+
+  auto control = getControlFromSequenceAsTwist(plan.header.stamp);
+
+  if (settings_.shift_control_sequence) {
+    shiftControlSequence();
   }
 
-  auto control = getControlFromSequenceAsTwist(
-    settings_.control_sequence_shift_offset_,
-    plan.header.stamp);
+  fallback(critics_data_.fail_flag);
 
-  shiftControlSequence();
   return control;
 }
 
-void Optimizer::shiftControlSequence()
+void Optimizer::fallback(bool fail)
```

Comment:
fallback doesn't try to compute again, that's a problem. We *cannot* return a control that is invalid.

Reply:
Our options are
- We fail to compute, so we reset & throw
- We fail to compute. so we reset and try again. If then we fail after `N` times, we throw. But this must happen in the same cycle.

Reply:
Oh. You mean after first failure reset then invoke again full pipeline without passing cmd_vel, got it

Reply:
Yeah :-)

Also quick question, does it make sense to ever retry (e.g. `reset()` and try again in the same cycle) more than 1 time? I see you added a parameter that people could do it `N` times. Is it actually useful to do it `N` times, or just once?

Reply:
Probably fixed

Reply:
Hm.. I guess it make sense to have either one, or infinite (not fail at all and try to recover internally). I don't think this parameter something that brings much cognitive complexity to user, but sure, it could using constantly 1, until some user says that he need more/infinite

Reply:
I think someone might need it eventually, to let optimizer more chances to restore by itself

May be just let it as is ? I don't think this option brings much cognitive complexity

Reply:
Agreed